### PR TITLE
cmake: dts: import devicetree symbols into CMake

### PR DESCRIPTION
**Pull request by tejlmand**
This commit introduces devicetree API in CMake so that devicetree
properties and register block can be used in the CMake build system.

The script scripts/dts/dts_cmake.py processes the edt.pickle and
generates a corresponding devicetree property map in a devicetree_target
that is then used in CMake.

The following devicetree API has been made available in Zephyr CMake:
```
- dt_nodelabel(<var> LABEL <label>)
- dt_node_exists(<var> NODE <node>)
- dt_node_has_status(<var> NODE <node> STATUS <status>)
- dt_prop(<var> NODE <node> PROPERTY <prop>)
- dt_prop(<var> NODE <node> INDEX <idx> PROPERTY <prop>)
- dt_num_regs(<var> NODE <node>)
- dt_reg_addr(<var> NODE <node> [INDEX <idx>])
- dt_reg_size(<var> NODE <node> [INDEX <idx>])
- dt_has_chosen(<var> PROPERTY <prop>)
- dt_chosen(<var> PROPERTY <prop>)
```

Signed-off-by: Marti Bolivar <marti.bolivar@nordicsemi.no>
Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

-------

This is needed by:  #36140

### CODE REVIEW

File: cmake/extensions.cmake
Comment (galak) [inline cmake/extensions.cmake:2267]: can we split this out into its own file.

Comment (tejlmand) [reply cmake/extensions.cmake:2267]: we could, but are we sure that we want to do that ?

Today, whenever one sees a CMake function or macro in use inside the that is not vanilla CMake, then there is one place to look, and that is `cmake/extensions.cmake`.

That makes it easy to lookup extension functions.

Only exception are areas where functions are used internally, such as: `cmake/flash/CMakeLists.txt` but those functions are not use outside that file, afaik.

We could of course establish a new scheme, like:
- `cmake/extensions/dt-extensions.cmake`
- `cmake/extensions/library-extensions.cmake`
  - for library related extensions, like `zephyr_library()`,  `zephyr_library_sources`, `zephyr_sources`, ...
- `cmake/extensions/generic-extensions.cmake`
  - for generic CMake functions with a twist, like `target_sources_ifdef()`, `add_subdirectory_ifdef()`.
- `cmake/extensions/misc-extensions.cmake`
  - There always needs to be a misc.

but doing such split (which would be outside the scope of this PR) will sometimes make it harder to know where a given function belongs (most functions will be easy, but not all).

Having only devicetree functions in a separate file just makes it inconsistent where functions are placed, imho.

If it is because you believe this feature could be reusable outside Zephyr scope, like with linker generator, then I see the point and a use-case.

Comment (galak) [reply cmake/extensions.cmake:2267]: the thought was that this could be usable outside of zephyr.

**Comment by galak** [comment]
How do we feel about use of the to select files to build?

For example something like:

`drivers/gpio/Kconfig.psoc6` that defines `GPIO_PSOC6`

and drivers/gpio/CMakeLists.txt could go from:

`zephyr_library_sources_ifdef(CONFIG_GPIO_PSOC6      gpio_psoc6.c)`

to

`zephyr_library_sources_dt_compat_enabled("cypress,psoc6-gpio"   gpio_psoc6.c)`

**Comment by galak** [comment]
Is this used for anything other than `zephyr,dtcm` and `zephyr,itcm` in the linker PR?

**Comment by tejlmand** [comment]
> Is this used for anything other than `zephyr,dtcm` and `zephyr,itcm` in the linker PR?

also here in linker PR:
https://github.com/zephyrproject-rtos/zephyr/blob/aa4f08e1220b1c5f53ad5b8bff87ae89730286e7/cmake/extensions.cmake#L2818-L2836
which is then used here:
https://github.com/zephyrproject-rtos/zephyr/blob/aa4f08e1220b1c5f53ad5b8bff87ae89730286e7/cmake/linker_script/arm/linker.cmake#L40-L49

but besides that, there have previous been requests regarding devicetree symbols in CMake.

**Comment by tejlmand** [comment]
> How do we feel about use of the to select files to build?

Personal I think that would cleanup the code not having the intermediate Kconfig setting for enabling the driver, but when looking at the Kconfig:
https://github.com/zephyrproject-rtos/zephyr/blob/33d5be865653dbc3628a86d1447fca8de6ce398e/drivers/gpio/Kconfig.psoc6#L7-L9 then the symbol does have a prompt, although with `default y`

So today it is possible to have `dt_compat_enabled` for cypress,psoc6-gpio and `CONFIG_GPIO_PSOC6=n` which would disable the sw-driver.
I don't know the exact usecase for that, but it does allow users to provide their own driver which would not be possible if we include the driver based directly on devicetree properties.

File: cmake/extensions.cmake
Comment (tejlmand) [reply cmake/extensions.cmake:2267]: > the thought was that this could be usable outside of zephyr.

makes sense, but those functions are useless without the generated `<build>/zephyr/dts.cmake` file which is created here https://github.com/zephyrproject-rtos/zephyr/blob/aecb7fc1b488ac3d017447d324b1e78e793f7620/scripts/dts/dts_cmake.py based on the `edt.pickle` file, https://github.com/zephyrproject-rtos/zephyr/blob/aecb7fc1b488ac3d017447d324b1e78e793f7620/scripts/dts/dts_cmake.py#L34

The `edt.pickle` file is created in the `gen_defines.py` script here: https://github.com/zephyrproject-rtos/zephyr/blob/33d5be865653dbc3628a86d1447fca8de6ce398e/scripts/dts/gen_defines.py#L929

so just placing those dt_* functions in its own CMake file doesn't really make them reusable.

If we really want to make it easier to reuse the devicetree --> CMake infrastructure outside Zephyr, then perhaps we should see if this could be streamlined with reusability in mind for all the parts involved:
- [\<zephyr\>/cmake/dts.cmake](https://github.com/zephyrproject-rtos/zephyr/blob/aecb7fc1b488ac3d017447d324b1e78e793f7620/cmake/dts.cmake)
- [\<zephyr\>/scripts/dts/dts_cmake.py](https://github.com/zephyrproject-rtos/zephyr/blob/aecb7fc1b488ac3d017447d324b1e78e793f7620/scripts/dts/dts_cmake.py)
- [\<zephyr\>/scripts/dts/gen_defines.py](https://github.com/zephyrproject-rtos/zephyr/blob/aecb7fc1b488ac3d017447d324b1e78e793f7620/scripts/dts/gen_defines.py)
- [\<zephyr\>/cmake/extensions.cmake](https://github.com/zephyrproject-rtos/zephyr/blob/aecb7fc1b488ac3d017447d324b1e78e793f7620/cmake/extensions.cmake), devicetree part only.

That sounds very reasonable to me, but I think it would be outside the scope of this PR.

Comment (mbolivar-nordic) [reply cmake/extensions.cmake:2267]: I agree with @tejlmand . If someone wants to use it outside of Zephyr they can pick up the patch at that point and make it do what they need. I'm not sure it's really necessary for this PR.

**Comment by mbolivar-nordic** [comment]
> I don't know the exact usecase for that, but it does allow users to provide their own driver which would not be possible if we include the driver based directly on devicetree properties.

This is important to users; see e.g. https://github.com/zephyrproject-rtos/zephyr/issues/10621#issuecomment-618834351 from @henrik

**Review by mbolivar-nordic — COMMENTED** [review]
As some background, the devicetree.h API in C uses "node identifier" as a precise term for the magic macros we use to refer to nodes, but we only have those because we cannot refer to a node by its path.

However, in CMake, we can refer to nodes by their paths, so I don't think the term "node identifier" should be used in this context. I would rather preserve the term "node identifier" for just the magic macro you see in C. That will be less ambiguous. In CMake, we can just take a path or alias when the user wants to tell us what node they want instead.

**Comment by mbolivar-nordic** [comment]
> How do we feel about use of the to select files to build?

Deciding whether or not to enable an entire driver should be left user-settable via Kconfig as I mentioned earlier.

However there are use cases, like here: https://github.com/zephyrproject-rtos/zephyr/blob/190532bcc4d54879c71a83a13382718630c21a70/drivers/sensor/bme280/bme280_i2c.c#L15

When `CONFIG_BME280=y`, I still only want to build bme280_i2c.c if there is an instance of the compatible on an I2C bus.

So if I could do something like this:

```cmake
if (CONFIG_BME280)
  dt_compat_on_bus_status_okay(result COMPATIBLE bosch,bme280 BUS i2c)
  if (result)
    zephyr_library_sources(bme280_i2c.c)
  endif()
  dt_compat_on_bus_status_okay(result COMPATIBLE bosch,bme280 BUS i2c)
  if (result)
    zephyr_library_sources(bme280_i2c.c)
  endif()
endif()
```

Then I think that would be a little cleaner than the current ifdef in the driver source.

**Comment by tejlmand** [comment]
> However, in CMake, we can refer to nodes by their paths, so I don't think the term "node identifier" should be used in this context. I would rather preserve the term "node identifier" for just the magic macro you see in C. That will be less ambiguous. In CMake, we can just take a path or alias when the user wants to tell us what node they want instead.

Good point, changed `node identifier` --> `node path`.

**Comment by tejlmand** [comment]
> Then I think that would be a little cleaner than the current ifdef in the driver source.
> 
> Edit: or even something like this:

Agree, let's look into the cleanest way to do this after the basic support for devicetree has been merged (This PR).
Initial feeling is that a 40 chars long function call is a bit long,  and because you would end-up line wrapping anyway, then this might be just as clean, imho:
```
dt_compat_on_bus_status_okay(result COMPATIBLE bosch,bme280 BUS i2c)
zephyr_library_sources_ifdef(result bme280_i2c.c)
```
Remember, we already have the `zephyr_library_sources_ifdef()` function.

File: cmake/extensions.cmake
Comment (galak) [inline cmake/extensions.cmake:2278]: should this be `NODELABEL <nodelabel>`?

File: cmake/extensions.cmake
Comment (galak) [inline cmake/extensions.cmake:2312]: should this be `dt_path_exists(<var> PATH <path>)`

File: cmake/extensions.cmake
Comment (galak) [inline cmake/extensions.cmake:2319]: What is returned in <var> if the node exists, what is returned if the node doesn't exist

File: cmake/extensions.cmake
Comment (galak) [inline cmake/extensions.cmake:2283]: what is returned if <label> doesn't exist?

File: cmake/extensions.cmake
Comment (galak) [inline cmake/extensions.cmake:2405]: what is `<prop>` doesnt exist, what is returned?

File: cmake/extensions.cmake
Comment (galak) [inline cmake/extensions.cmake:2563]: Seems like we should have a `dt_has_prop` if needed and implement this using that.

File: cmake/extensions.cmake
Comment (galak) [inline cmake/extensions.cmake:2608]: Is there a reason this just doesn't call `dt_prop`?

**Review by galak — CHANGES_REQUESTED** [review]
See inline comments:

General comments:
* cleanup docs to clarify return values for true/false cases as well as on "errors" (ie path doesn't exist, prop doesn't exist, etc)
* replace NODE with PATH

File: cmake/dts.cmake
Comment (mbolivar-nordic) [inline cmake/dts.cmake:265]:
```suggestion
    message(STATUS "Including generated dts.cmake file: ${DTS_CMAKE}")
```

This makes it clearer to the user what its purpose is IMO.

File: cmake/extensions.cmake
Comment (mbolivar-nordic) [reply cmake/extensions.cmake:2278]: > should this be `NODELABEL <nodelabel>`?

I think so too.

File: cmake/extensions.cmake
Comment (mbolivar-nordic) [inline cmake/extensions.cmake:2303]: Which means this target property needs updating too.

File: cmake/extensions.cmake
Comment (mbolivar-nordic) [inline cmake/extensions.cmake:2305]: I kind of want this to be a fatal error at first, with a possibility to relax it later by adding a new `DEFAULT my-default-value` argument to req_single_args.

Is there a good reason not to do it that way?

The kconfigfunctions.py functions that return paths do return empty strings when there's no such node, but that is because we have to have *something* to return since we parse everything in the entire kconfig tree and sometimes we're asking about nodes that only exist on a different SoC. So I don't think they're a good example to follow here.

File: cmake/extensions.cmake
Comment (mbolivar-nordic) [reply cmake/extensions.cmake:2312]: I think `dt_node_exists` is a good name for the function. It should take `PATH <path>` for now, but we can extend it later to also accept `ALIAS <alias>`, `NODELABEL <label>`, etc. without changing the name. I think that would be nicer than having separate `dt_alias_exists` ,`dt_nodelabel_exists`, etc.

File: cmake/extensions.cmake
Comment (mbolivar-nordic) [inline cmake/extensions.cmake:2360]: I think this should be `PATH <path>` for now; we can extend it to `ALIAS`, etc. later. Similar comments to other functions.

File: scripts/dts/dts_cmake.py
Comment (mbolivar-nordic) [inline scripts/dts/dts_cmake.py:34]: This is a weird nit, but I try to avoid assuming CPython like this when I can. Lines like this work in CPython because it's got a refcount-based garbage collector, so as soon as the reference to the open file gets dropped on the next line, it gets closed. But other Pythons don't have that behavior, so it's cleaner to use a context manager.

```py
with open(args.edt_pickle), 'rb') as f:
    edt = pickle.load(f)
```

This is the pattern we use everywhere else in zephyr.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
I generally agree with Kumar's cleanup suggestions, with a couple of other ideas and additional comments inline. Thanks.

File: cmake/extensions.cmake
Comment (tejlmand) [reply cmake/extensions.cmake:2305]: to allow code to just ask for a node with a given label.

If no node is having that label, then there is nothing to return.

If someone asks for `some-node`, then how should they know prior to their call if there is any node with such label ?
And how will a fatal error help them.

By undefining the return var, they can do:
```
dt_label(node LABEL some-node)
if(DEFINED node)
  # Node exists, do something related to that node
endif()
```

File: cmake/extensions.cmake
Comment (tejlmand) [reply cmake/extensions.cmake:2563]: Do you want to ask whether a given node has a particular property ?
Like this:
```
dt_has_prop(result PATH <path> PROPERTY <prop>)
if(${result})
  dt_prop(value PATH <path> PROPERTY <prop>)
endif()
```

what's the value in the double call, the `dt_prop()` gives you everything in a single call.
```
dt_prop(value PATH <path> PROPERTY <prop>)
if(NOT DEFINED value)
  message("Node: ${path} does not have property ${prop}")
endif()

if(DEFINED value)
  message("Node: ${path} has property ${prop}, value is: ${value}")
endif()
```
with CMake, the return variable is `undefined` if the property doesn't exists for that node.
And if the property exists, the value of that property is returned.

So I really don't see the value in `dt_has_prop()`

Comment (tejlmand) [reply cmake/extensions.cmake:2563]: and of course if the property value is FALSE, test like:
```
dt_prop(value PATH <path> PROPERTY <prop>)
if(DEFINED value AND ${value})
  message("Node: ${path} has property ${prop}, and is TRUE")
endif()

if(DEFINED value AND NOT ${value})
  message("Node: ${path} has property ${prop}, and is FALSE")
endif()
```

File: cmake/extensions.cmake
Comment (tejlmand) [reply cmake/extensions.cmake:2283]: the the return variable will be undefined, and you can test it like:
```
dt_nodelabel(nodepath LABEL mylabel)
if(DEFINED nodepath)
  # I got a node path, do something
endif()
```

File: cmake/extensions.cmake
Comment (tejlmand) [reply cmake/extensions.cmake:2319]: Every node in the system will have the `DT_NODE|<node-path>` value TRUE, see https://github.com/zephyrproject-rtos/zephyr/blob/a3400ceba261351b435174f8980b9447e3dfea09/scripts/dts/dts_cmake.py#L42
meaning that this line will return TRUE if the node exists:
https://github.com/zephyrproject-rtos/zephyr/blob/a3400ceba261351b435174f8980b9447e3dfea09/cmake/extensions.cmake#L2342
but if the node did not exist, then `FALSE` is returned here:
https://github.com/zephyrproject-rtos/zephyr/blob/a3400ceba261351b435174f8980b9447e3dfea09/cmake/extensions.cmake#L2344

So answer is:
- TRUE when the node exists
- FALSE otherwise

File: cmake/extensions.cmake
Comment (tejlmand) [reply cmake/extensions.cmake:2405]: Then the `<var>` will be undefined.

If it exists, then `<var>` is defined with the value of the property.

File: cmake/extensions.cmake
Comment (tejlmand) [reply cmake/extensions.cmake:2608]: Only reason is the chosen nodes have their own property names like this: `"DT_CHOSEN|${DT_CHOSEN_PROPERTY}"` but `dt_prop()` could easily be extended to taken an extra argument, but initially I just tried to stay close to existing API.

File: cmake/extensions.cmake
Comment (mbolivar-nordic) [reply cmake/extensions.cmake:2305]: > If someone asks for `some-node`, then how should they know prior to their call if there is any node with such label ?

With a `dt_node_exists` helper, of course :)

Comment (mbolivar-nordic) [reply cmake/extensions.cmake:2305]: ... or a `DEFAULT ""` extra keyword argument to the function, since the empty string evaluates to false.

File: cmake/extensions.cmake
Comment (mbolivar-nordic) [reply cmake/extensions.cmake:2319]: > So answer is:

I think @galak is asking you to document the behavior

File: cmake/extensions.cmake
Comment (mbolivar-nordic) [reply cmake/extensions.cmake:2283]: > the the return variable will be undefined, and you can test it like:

Same here, I think the request is for documented behavior.

File: cmake/extensions.cmake
Comment (mbolivar-nordic) [reply cmake/extensions.cmake:2405]: > Then the `<var>` will be undefined.

Docs?

File: cmake/extensions.cmake
Comment (tejlmand) [reply cmake/extensions.cmake:2319]: I felt he was asking also because he was unsure, but anything that is not clear in the description of the function itself should of course also be improved.

I just wanted to give a detailed explanation here first, before updating the PR in general with all requested changes.

File: cmake/extensions.cmake
Comment (tejlmand) [reply cmake/extensions.cmake:2305]: > With a `dt_node_exists` helper, of course :)

Could you explain a bit on how such a flow would look, cause i'm not completely sure I understand your intention.
I want to find the node with nodelabel "something", then how can I ask whether that node exists before getting it ?
```
dt_node_exists(result PATH ${node_path})  # <---  but I don't know the path, as that is what I will be asking for in next line
dt_label(node_path NODELABEL "something") # <-- Here I get the path.
```

Or did you mean that `dt_node_exists()` should also take node label as argument, like this:
```
dt_node_exists(result NODELABEL "something")
```

but in such case we'll end up with constructs like this:
```
dt_node_exists(result NODELABEL "something")
if(result)
  dt_label(node_path NODELABEL "something")
  # Now do something with that node, like
  # dt_prop(result PATH ${node_path})
endif()
```

but why is that better than:
```
dt_label(node_path NODELABEL "something")
if(DEFINED node_path)
  # Now do something with that node, like
  # dt_prop(result PATH ${node_path})
endif()
```
In CMake we can use the knowledge of whether the variable is undefined versus an empty string / TRUE / FALSE / value.

So in current PR, I have both the knowledge of whether or not the node exists, but also the knowledge of its node path if it does exists, so why split that into two function calls ?
What's the exact gain.

File: cmake/extensions.cmake
Comment (tejlmand) [reply cmake/extensions.cmake:2360]: Changed all `NODE <path>` to `PATH <path>`.

File: cmake/extensions.cmake
Comment (tejlmand) [reply cmake/extensions.cmake:2305]: > .. or a `DEFAULT ""` extra keyword argument to the function, since the empty string evaluates to false.

but we are not returning an empty string.
We are undefining the variable.

What would be the exact gain of returning some default when the node is not there ?

Comment (tejlmand) [reply cmake/extensions.cmake:2305]: Small example of the difference between undefined and an empty string:
```
set(TEST_VAR)

if(DEFINED TEST_VAR)
  message("TEST_VAR is defined, value=${TEST_VAR}")
else()
  message("TEST_VAR is undefined")
endif()

set(TEST_VAR "")

if(DEFINED TEST_VAR)
  message("TEST_VAR is defined, value=${TEST_VAR}")
else()
  message("TEST_VAR is undefined")
endif()
```
results in:
```
$ cmake -P z.cmake
TEST_VAR is undefined
TEST_VAR is defined, value=
```
So the `if(DEFINED <var>)` returns true, even for an empty string, but false if the variable is undefined.
```
set(MY_VAR)    # <-- This undefines the var
set(MY_VAR "") # <-- This defines the var to the empty string.
```

File: scripts/dts/dts_cmake.py
Comment (tejlmand) [reply scripts/dts/dts_cmake.py]: Fixed

File: cmake/extensions.cmake
Comment (tejlmand) [reply cmake/extensions.cmake]: Fixed

File: cmake/extensions.cmake
Comment (tejlmand) [reply cmake/extensions.cmake:2322]: Function description improved

File: cmake/extensions.cmake
Comment (tejlmand) [reply cmake/extensions.cmake:2283]: Function description improved

File: cmake/extensions.cmake
Comment (tejlmand) [reply cmake/extensions.cmake]: Function description improved

File: cmake/dts.cmake
Comment (tejlmand) [reply cmake/dts.cmake]: fixed

File: cmake/extensions.cmake
Comment (tejlmand) [reply cmake/extensions.cmake]: fixed

File: cmake/extensions.cmake
Comment (mbolivar-nordic) [reply cmake/extensions.cmake:2308]: OK, I am convinced. Thanks.

File: cmake/dts.cmake
Comment (mbolivar-nordic) [inline cmake/dts.cmake:30]: On second thought, can we name this gen_dts_cmake.py, to make it more obvious what it does?

File: cmake/dts.cmake
Comment (mbolivar-nordic) [inline cmake/dts.cmake:263]: gen_dts_cmake.py

Comment (tejlmand) [reply cmake/dts.cmake]: fixed

File: cmake/dts.cmake
Comment (tejlmand) [reply cmake/dts.cmake]: fixed

**Comment by mbolivar-nordic** [comment]
@tejlmand I pushed a fixup branch here:

https://github.com/mbolivar-nordic/zephyr/tree/dts_cmake

Please fetch, review, and squash if it looks OK with you. See commit logs for justifications.

File: cmake/extensions.cmake
Comment (mbolivar-nordic) [reply cmake/extensions.cmake:2602]: After thinking and discussing with @tejlmand, I'm resolving this "won't do" because this can't be implemented using dt_has_prop() with the current edtlib.

The reason why is that edtlib.Node.props only captures properties that are described in bindings. Since /chosen has no binding, it's not covered, and you have to use special-purpose edtlib APIs for /chosen. This is the case in C also; `DT_HAS_CHOSEN(prop)` in C is defined with a totally different set of generated macros.

File: cmake/extensions.cmake
Comment (mbolivar-nordic) [reply cmake/extensions.cmake:2649]: > Is there a reason this just doesn't call `dt_prop`?

The reason here is the same as discussed in the same thread below on dt_has_chosen(): edtlib treats /chosen specially because it has no binding; it doesn't work like "regular" properties with bindings.

**Comment by mbolivar-nordic** [comment]
> If we're on the same page about all of those, I'm +1 to get this in.

And if I've made any terrible mistakes, please let me know :smile:

**Comment by tejlmand** [comment]
> 
> 
> > If we're on the same page about all of those, I'm +1 to get this in.
> 
> And if I've made any terrible mistakes, please let me know 😄

Not a mistake, but placed a comment here: https://github.com/mbolivar-nordic/zephyr/commit/d3b44e2978f5d9fb234d533fd8369ce6bb9d5ca2#r55395442
which allows us to handle the edge case without a lot of extra code.

**Comment by tejlmand** [comment]
> Please fetch, review, and squash if it looks OK with you. See commit logs for justifications.

Thanks a lot, I have taken in your changes here: https://github.com/zephyrproject-rtos/zephyr/compare/f0cae1d28edafda85345fe24de6e37dc64c0063c..736dc0b1edc516de74568881ca44c7f484e0a79c

and then made a small update according to my comments here:
https://github.com/zephyrproject-rtos/zephyr/compare/736dc0b1edc516de74568881ca44c7f484e0a79c..7dbfc0094d73a90822d69041454792573cce6e0a

**Review by mbolivar-nordic — APPROVED** [review]
Looks good now, thanks.

**Comment by mbolivar-nordic** [comment]
I think we should dismiss @galak's review as "addressed" to get this in for feature freeze.

**Review by nashif — APPROVED** [review]

**Pull request merged**
